pmlc/dialect/abi/ir/ops.td

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Intel Corporation
+
+#ifndef __PML_ABI_OPS__
+#define __PML_ABI_OPS__
+
+include "mlir/Dialect/StandardOps/IR/StandardOpsBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+include "pmlc/util/enums.td"
+
+def ABI_Dialect : Dialect {
+  let name = "abi";
+  let cppNamespace = "pmlc::dialect::abi";
+}
+
+class ABI_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<ABI_Dialect, mnemonic, traits>;
+
+def ABI_LoopOp : ABI_Op<"loop", [
+  IsolatedFromAbove, HasParent<"ModuleOp">, RecursiveSideEffects]> {
+  let summary = "Repeat runs of a network";
+  let description = [{
+    The "abi.loop" operation represents repeated runs of a program,
+    taking new inputs and generating new outputs each time.
+    This allows loop-invariant code motion to move setup and teardown
+    operations outside of the loop, so that their cost is ammortized over
+    many runs of the program.
+  }];
+  let regions = (region AnyRegion:$initRegion, AnyRegion:$bodyRegion, AnyRegion:$finiRegion);
+  let extraClassDeclaration = [{
+    static mlir::StringRef getNetworkFieldTypesAttrName() { return "networkFieldTypes"; }
+
+    std::vector<mlir::Type> getNetworkFieldTypes() {
+      std::vector<mlir::Type> result;
+      auto arrayAttr = getAttrOfType<mlir::ArrayAttr>(getNetworkFieldTypesAttrName());
+      if (arrayAttr) {
+        for (auto attr : arrayAttr) {
+          if (auto tyAttr = attr.dyn_cast<mlir::TypeAttr>()) {
+            result.emplace_back(tyAttr.getValue());
+          }
+        }
+      }
+      return result;
```

Comment:
Please move C++ definitions to a `.cc` file, we'd like to minimize the amount of C++ code in `.td` files.

Reply:
Will do

Reply:
Done